File: previous_code/functions/src/services/LeakScraper.ts
import axios from 'axios';
import fs from 'fs-extra';
import path from 'path';
import { URL } from 'url';
import { createObjectCsvWriter } from 'csv-writer';

interface SearchResult {
  title: string;
  url: string;
  snippet: string;
  query: string;
  page: number;
  date: string;
}

class LeakScraper {
  private creatorName: string;
  private apiKey: string;
  private searchEngineId: string;
  private maxSearches: number;
  private apiCalls: number;
  private baseDir: string;
  private outputFile: string;
  private uniqueUrls: Set<string>;

  constructor(creatorName: string, apiKey: string, searchEngineId: string, maxSearches: number = 100) {
    this.creatorName = creatorName;
    this.apiKey = apiKey;
    this.searchEngineId = searchEngineId;
    this.maxSearches = maxSearches;
    this.apiCalls = 0;

    // Create output directory
    this.baseDir = path.join(process.cwd(), '..', 'data', 'leak_detection_results');
    fs.ensureDirSync(this.baseDir);

    // Default output file
    this.outputFile = path.join(this.baseDir, `${creatorName.replace(/\s+/g, '_')}_results.csv`);

    // Track URLs to avoid duplicates
    this.uniqueUrls = new Set<string>();

    // Load existing results if file exists
    this.loadExistingResults();
  }

  private loadExistingResults(): void {
    try {
      if (fs.existsSync(this.outputFile)) {
        const content = fs.readFileSync(this.outputFile, 'utf8');
        const lines = content.split('\n').slice(1); // Skip header
        
        lines.forEach(line => {
          if (line) {
            const parts = line.split(',');
            if (parts.length > 1) {
              const url = parts[1].replace(/"/g, '');
              this.uniqueUrls.add(url);
            }
          }
        });
        
        console.log(`✅ Loaded ${this.uniqueUrls.size} previously found URLs`);
      }
    } catch (error) {
      console.log(`ℹ️ No previous results loaded: ${error}`);
    }
  }

  private adaptiveBatchKeywords(keywords: string[], maxBatchSize: number = 1): string[][] {
    // Group keywords into batches
    return keywords.reduce((batches: string[][], keyword, index) => {
      const batchIndex = Math.floor(index / maxBatchSize);
      
      if (!batches[batchIndex]) {
        batches[batchIndex] = [];
      }
      
      batches[batchIndex].push(keyword);
      return batches;
    }, []);
  }

  private buildQuery(keywordBatch: string[]): string {
    const queryParts = keywordBatch.map(keyword => {
      // Check if the keyword already has quotes
      if (keyword.startsWith('"') && keyword.endsWith('"')) {
        return keyword;
      } else if (keyword.includes('site:')) {
        // Don't add quotes to site-specific searches
        return keyword;
      } else {
        // Add the keyword as is
        return keyword;
      }
    });

    // Join with OR operator
    return queryParts.join(' OR ');
  }

  private getDateRestrict(timeframe: string): string {
    if (timeframe === 'today') {
      return 'd1';
    } else if (timeframe.includes('days')) {
      const days = parseInt(timeframe.split(' ')[1], 10);
      return `d${days}`;
    } else if (timeframe.includes('weeks')) {
      const weeks = parseInt(timeframe.split(' ')[1], 10);
      const days = weeks * 7;
      return `d${days}`;
    } else if (timeframe.includes('months')) {
      const months = parseInt(timeframe.split(' ')[1], 10);
      const days = months * 30;
      return `d${days}`;
    } else {
      // Default to last 30 days
      return 'd30';
    }
  }

  private async searchBatch(keywordBatch: string[], dateRestrict: string, maxPages: number = 10): Promise<SearchResult[]> {
    // Build combined query
    const query = this.buildQuery(keywordBatch);
    const encodedQuery = encodeURIComponent(query);

    console.log(`\n🔍 Searching batch: ${keywordBatch.slice(0, 3).join(' | ')}...`);
    if (keywordBatch.length > 3) {
      console.log(`   ...and ${keywordBatch.length - 3} more keywords`);
    }
    console.log(`   Query: ${query.length > 100 ? query.substring(0, 100) + '...' : query}`);

    // Store results
    const batchResults: SearchResult[] = [];

    // Search with pagination
    for (let page = 1; page <= maxPages; page++) {
      // Check if we've hit our search limit
      if (this.apiCalls >= this.maxSearches) {
        console.log(`⚠️ Reached search limit (${this.apiCalls}/${this.maxSearches})`);
        return batchResults;
      }

      // Calculate start index for pagination
      const startIndex = ((page - 1) * 10) + 1;

      // Build URL parameters
      const urlParams = [
        `q=${encodedQuery}`,
        `cx=${this.searchEngineId}`,
        `key=${this.apiKey}`,
        'num=10', // Always 10 (API limit)
        `start=${startIndex}`
      ];

      // Add date restriction
      urlParams.push(`dateRestrict=${dateRestrict}`);

      // ALWAYS add exactTerms to ensure creator name is present
      urlParams.push(`exactTerms=${encodeURIComponent(this.creatorName)}`);

      // Build final URL
      const apiUrl = `https://www.googleapis.com/customsearch/v1?${urlParams.join('&')}`;

      try {
        console.log(`   📄 Page ${page}/${maxPages}...`);
        const response = await axios.get(apiUrl);
        this.apiCalls += 1;
        const data = response.data;

        // Handle API errors
        if (data.error) {
          const errorMsg = data.error.message || 'Unknown error';
          console.log(`   ⚠️ API error: ${errorMsg}`);

          // Check for quota exceeded errors
          if (errorMsg.toLowerCase().includes('quota')) {
            console.log('❌ API quota exceeded! Stopping searches.');
            return batchResults;
          }

          await new Promise(resolve => setTimeout(resolve, 2000)); // Wait longer if we hit an error
          continue;
        }

        if (data.items) {
          const resultCount = data.items.length;
          let newCount = 0;
          console.log(`\n   📋 Page ${page} Results:`);

          for (let i = 0; i < resultCount; i++) {
            const item = data.items[i];
            const title = item.title || '';
            const link = item.link || '';
            const snippet = item.snippet || '';

            // Show all links in terminal
            const isNew = !this.uniqueUrls.has(link);
            const status = isNew ? '🆕' : '📎';
            console.log(`   ${status} ${i + 1}. ${title.substring(0, 50)}... - ${link}`);

            // Only add if it's a new URL
            if (isNew) {
              batchResults.push({
                title,
                url: link,
                snippet,
                query: JSON.stringify(keywordBatch),
                page,
                date: new Date().toISOString().split('T')[0]
              });
              this.uniqueUrls.add(link);
              newCount += 1;
            }
          }

          console.log(`\n   ✅ Found ${resultCount} results, ${newCount} new URLs`);

          // If we do not find new results on a new page
          if (resultCount === 0) {
            console.log('   ℹ️ No more results available');
            break;
          }
        } else {
          console.log('   ⚠️ No results found on this page');
          break; // No need to check more pages
        }

        // Respect API rate limits
        await new Promise(resolve => setTimeout(resolve, 2000));
      } catch (error: any) {
        console.log(`   ⚠️ Error: ${error.message}`);
        await new Promise(resolve => setTimeout(resolve, 2000));
      }
    }

    return batchResults;
  }

  public async runScan(keywords: string[], timeframe: string, maxSearches?: number): Promise<SearchResult[]> {
    if (maxSearches !== undefined) {
      this.maxSearches = maxSearches;
    }

    console.log(`\n🚀 Starting leak scan for: ${this.creatorName}`);
    console.log(`📅 Timeframe: ${timeframe}`);
    console.log(`🔢 Search budget: ${this.maxSearches} API calls`);
    console.log(`🔍 Using ${keywords.length} keywords: ${keywords}`);

    // Reset API calls counter
    this.apiCalls = 0;

    // Convert timeframe to date_restrict parameter
    const dateRestrict = this.getDateRestrict(timeframe);
    console.log(`ℹ️ Converted timeframe to date parameter: ${dateRestrict}`);

    // Group keywords into appropriately sized batches
    const batches = this.adaptiveBatchKeywords(keywords);
    console.log(`ℹ️ Created ${batches.length} batches of keywords`);

    // Store all results
    const allResults: SearchResult[] = [];

    // Process each batch
    for (let batchNum = 0; batchNum < batches.length; batchNum++) {
      // Check if we've hit our search limit
      if (this.apiCalls >= this.maxSearches) {
        console.log(`⚠️ Reached search limit (${this.apiCalls}/${this.maxSearches})`);
        break;
      }

      console.log(`\n🔍 Processing batch ${batchNum + 1}/${batches.length}`);

      // Search this batch with pagination
      const batchResults = await this.searchBatch(batches[batchNum], dateRestrict);
      allResults.push(...batchResults);

      // Show progress
      console.log(`Progress: ${this.apiCalls}/${this.maxSearches} API calls used (${(this.apiCalls / this.maxSearches * 100).toFixed(1)}%)`);
      console.log(`Found ${allResults.length} unique URLs so far`);

      // Save results after each batch
      await this.saveResults(allResults);
    }

    console.log(`\n✅ Scan complete! Results saved to ${this.outputFile}`);
    console.log(`🔢 API calls used: ${this.apiCalls}/${this.maxSearches}`);
    console.log(`🔗 Unique URLs found: ${allResults.length}`);

    // Print summary of top domains
    this.printDomainSummary(allResults);

    return allResults;
  }

  private printDomainSummary(results: SearchResult[]): void {
    if (!results.length) {
      return;
    }

    // Extract domains from URLs
    const domains: string[] = [];
    results.forEach(result => {
      try {
        // Extract domain from URL
        const url = new URL(result.url);
        domains.push(url.hostname);
      } catch {
        // Skip invalid URLs
      }
    });

    // Count domains
    const domainCounts: Record<string, number> = {};
    domains.forEach(domain => {
      domainCounts[domain] = (domainCounts[domain] || 0) + 1;
    });

    // Sort by count
    const sortedDomains = Object.entries(domainCounts)
      .sort((a, b) => b[1] - a[1])
      .slice(0, 10);

    // Print top domains
    console.log('\n📊 Top Domains with Content:');
    sortedDomains.forEach(([domain, count], i) => {
      console.log(`   ${i + 1}. ${domain}: ${count} URLs`);
    });
  }

  public async saveResults(results: SearchResult[]): Promise<void> {
    if (!results.length) {
      console.log('⚠️ No results to save');
      return;
    }

    // Create CSV writer
    const csvWriter = createObjectCsvWriter({
      path: this.outputFile,
      header: [
        { id: 'title', title: 'title' },
        { id: 'url', title: 'url' },
        { id: 'snippet', title: 'snippet' },
        { id: 'query', title: 'query' },
        { id: 'page', title: 'page' },
        { id: 'date', title: 'date' }
      ]
    });

    await csvWriter.writeRecords(results);
    console.log(`✅ Saved ${results.length} results to ${this.outputFile}`);
  }

  public async saveTempResults(results: SearchResult[], creatorName: string): Promise<string | null> {
    if (!results.length) {
      return null;
    }

    const tempDir = path.join(process.cwd(), '..', 'data', 'temp_results');
    fs.ensureDirSync(tempDir);

    const timestamp = new Date().toISOString().replace(/[:.]/g, '').replace('T', '_').slice(0, 15);
    const tempFile = path.join(tempDir, `${creatorName.replace(/\s+/g, '_')}_temp_${timestamp}.csv`);

    // Create CSV writer
    const csvWriter = createObjectCsvWriter({
      path: tempFile,
      header: [
        { id: 'title', title: 'title' },
        { id: 'url', title: 'url' },
        { id: 'snippet', title: 'snippet' },
        { id: 'query', title: 'query' },
        { id: 'page', title: 'page' },
        { id: 'date', title: 'date' }
      ]
    });

    await csvWriter.writeRecords(results);
    console.log(`✅ Saved temp results to ${tempFile} for keyword learning`);

    return tempFile;
  }
}

export default LeakScraper;

File: previous_code/python/leak_scraper.py
import requests
import pandas as pd
import urllib.parse
import time
import json
import os
from datetime import datetime


class LeakScraper:
    def __init__(self, creator_name, api_key, search_engine_id, max_searches=100):
        """
        Initialize a leak scraper with adaptive batch sizing

        Args:
            creator_name: The creator to search for
            api_key: Google Custom Search API key
            search_engine_id: Google Custom Search Engine ID
            max_searches: Maximum API calls to make
        """
        self.creator_name = creator_name
        self.api_key = api_key
        self.search_engine_id = search_engine_id
        self.max_searches = max_searches
        self.api_calls = 0

        # Create output directory
        self.base_dir = os.path.join(os.getcwd(), "leak_detection_results")
        os.makedirs(self.base_dir, exist_ok=True)

        # Default output file
        self.output_file = os.path.join(self.base_dir, f"{creator_name.replace(' ', '_')}_results.csv")

        # Track URLs to avoid duplicates
        self.unique_urls = set()

        # Load existing results if file exists
        self.load_existing_results()

    def load_existing_results(self):
        """Load previously discovered URLs to avoid duplicates"""
        try:
            if os.path.exists(self.output_file):
                df = pd.read_csv(self.output_file)
                self.unique_urls = set(df['url'].tolist())
                print(f"✅ Loaded {len(self.unique_urls)} previously found URLs")
        except Exception as e:
            print(f"ℹ️ No previous results loaded: {e}")

    def adaptive_batch_keywords(self, keywords, max_batch_size=1):
        """Group keywords into batches with maximum 1 keywords per batch"""
        # Always create batches with max 2 keywords per batch
        return [keywords[i:i + max_batch_size] for i in range(0, len(keywords), max_batch_size)]

    def build_query(self, keyword_batch):
        """Build an optimized query from a batch of keywords"""
        query_parts = []

        for keyword in keyword_batch:
            # Check if the keyword already has quotes
            if keyword.startswith('"') and keyword.endswith('"'):
                query_parts.append(keyword)
            elif "site:" in keyword:
                # Don't add quotes to site-specific searches
                query_parts.append(keyword)
            else:
                # Add quotes for exact phrase matching
                query_parts.append(keyword)

        # Join with OR operator
        query = " OR ".join(query_parts)
        return query

    def get_date_restrict(self, timeframe):
        """Convert user timeframe to API date_restrict parameter"""
        if timeframe == "today":
            return "d1"
        elif timeframe == "lifetime":
            return ""  # No date restriction for all time
        elif "days" in timeframe:
            days = int(timeframe.split()[1])
            return f"d{days}"
        elif "weeks" in timeframe:
            weeks = int(timeframe.split()[1])
            days = weeks * 7
            return f"d{days}"
        elif "months" in timeframe:
            months = int(timeframe.split()[1])
            days = months * 30
            return f"d{days}"
        else:
            # Default to last 30 days
            return "d30"

    def search_batch(self, keyword_batch, date_restrict, max_pages=10):
        """Search for a batch of keywords with pagination"""
        # Build combined query
        query = self.build_query(keyword_batch)
        encoded_query = urllib.parse.quote(query)

        print(f"\n🔍 Searching batch: {' | '.join(keyword_batch[:3])}...")
        if len(keyword_batch) > 3:
            print(f"   ...and {len(keyword_batch) - 3} more keywords")
        print(f"   Query: {query[:100]}..." if len(query) > 100 else f"   Query: {query}")

        # Store results
        batch_results = []

        # Search with pagination
        for page in range(1, max_pages + 1):
            # Check if we've hit our search limit
            if self.api_calls >= self.max_searches:
                print(f"⚠️ Reached search limit ({self.api_calls}/{self.max_searches})")
                return batch_results

            # Calculate start index for pagination
            start_index = ((page - 1) * 10) + 1

            # Build URL parameters
            url_params = [
                f"q={encoded_query}",
                f"cx={self.search_engine_id}",
                f"key={self.api_key}",
                "num=10",  # Always 10 (API limit)
                f"start={start_index}"
            ]

            # Add date restriction only if specified (not empty for lifetime)
            if date_restrict:
                url_params.append(f"dateRestrict={date_restrict}")

            # ALWAYS add exactTerms to ensure creator name is present
            url_params.append(f"exactTerms={self.creator_name}")

            # Build final URL
            api_url = f"https://www.googleapis.com/customsearch/v1?{'&'.join(url_params)}"

            try:
                print(f"   📄 Page {page}/{max_pages}...")
                response = requests.get(api_url)
                self.api_calls += 1
                data = response.json()

                # Handle API errors
                if "error" in data:
                    error_msg = data["error"].get("message", "Unknown error")
                    print(f"   ⚠️ API error: {error_msg}")

                    # Check for quota exceeded errors
                    if "quota" in error_msg.lower():
                        print("❌ API quota exceeded! Stopping searches.")
                        return batch_results

                    time.sleep(2)  # Wait longer if we hit an error
                    continue

                if "items" in data:
                    result_count = len(data["items"])
                    new_count = 0
                    print(f"\n   📋 Page {page} Results:")

                    for i, item in enumerate(data["items"], 1):
                        title = item.get("title", "")
                        link = item.get("link", "")
                        snippet = item.get("snippet", "")

                        # Show all links in terminal
                        is_new = link not in self.unique_urls
                        status = "🆕" if is_new else "📎"
                        print(f"   {status} {i}. {title[:50]}... - {link}")

                        # Only add if it's a new URL
                        if is_new:
                            batch_results.append({
                                "title": title,
                                "url": link,
                                "snippet": snippet,
                                "query": str(keyword_batch),
                                "page": page,
                                "date": datetime.now().strftime('%Y-%m-%d')
                            })
                            self.unique_urls.add(link)
                            new_count += 1

                    print(f"\n   ✅ Found {result_count} results, {new_count} new URLs")

                    # If we do not find new results on a new page
                    if result_count == 0:
                        print("   ℹ️ No more results available")
                        break
                else:
                    print("   ⚠️ No results found on this page")
                    break  # No need to check more pages

                # Respect API rate limits
                time.sleep(2)

            except Exception as e:
                print(f"   ⚠️ Error: {e}")
                time.sleep(2)

        return batch_results

    def run_scan(self, keywords, timeframe, max_searches=None):
        """Run a scan with user-provided keywords and timeframe"""
        if max_searches is not None:
            self.max_searches = max_searches

        print(f"\n🚀 Starting leak scan for: {self.creator_name}")
        print(f"📅 Timeframe: {timeframe}")
        print(f"🔢 Search budget: {self.max_searches} API calls")
        print(f"🔍 Using {len(keywords)} keywords: {keywords}")

        # Reset API calls counter
        self.api_calls = 0

        # Convert timeframe to date_restrict parameter
        date_restrict = self.get_date_restrict(timeframe)
        print(f"ℹ️ Converted timeframe to date parameter: {date_restrict or 'No date restriction (All Time)'}")

        # Group keywords into appropriately sized batches
        batches = self.adaptive_batch_keywords(keywords)
        print(f"ℹ️ Created {len(batches)} batches of keywords")

        # Store all results
        all_results = []

        # Process each batch
        for batch_num, keyword_batch in enumerate(batches, 1):
            # Check if we've hit our search limit
            if self.api_calls >= self.max_searches:
                print(f"⚠️ Reached search limit ({self.api_calls}/{self.max_searches})")
                break

            print(f"\n🔍 Processing batch {batch_num}/{len(batches)}")

            # Search this batch with pagination
            batch_results = self.search_batch(keyword_batch, date_restrict)
            all_results.extend(batch_results)

            # Show progress
            print(
                f"Progress: {self.api_calls}/{self.max_searches} API calls used ({self.api_calls / self.max_searches * 100:.1f}%)")
            print(f"Found {len(all_results)} unique URLs so far")

            # Save results after each batch
            self.save_results(all_results)

        print(f"\n✅ Scan complete! Results saved to {self.output_file}")
        print(f"🔢 API calls used: {self.api_calls}/{self.max_searches}")
        print(f"🔗 Unique URLs found: {len(all_results)}")

        # Print summary of top domains
        self.print_domain_summary(all_results)

        return all_results

    def print_domain_summary(self, results):
        """Print a summary of the top domains found"""
        if not results:
            return

        # Extract domains from URLs
        domains = []
        for result in results:
            url = result['url']
            try:
                # Extract domain from URL
                domain = url.split('//')[1].split('/')[0]
                domains.append(domain)
            except:
                pass

        # Count domains
        domain_counts = {}
        for domain in domains:
            if domain in domain_counts:
                domain_counts[domain] += 1
            else:
                domain_counts[domain] = 1

        # Sort by count
        sorted_domains = sorted(domain_counts.items(), key=lambda x: x[1], reverse=True)

        # Print top domains
        print("\n📊 Top Domains with Content:")
        for i, (domain, count) in enumerate(sorted_domains[:10], 1):
            print(f"   {i}. {domain}: {count} URLs")

    def save_results(self, results):
        """Save results to CSV file"""
        if not results:
            print("⚠️ No results to save")
            return

        df = pd.DataFrame(results)
        df.to_csv(self.output_file, index=False)
        print(f"✅ Saved {len(results)} results to {self.output_file}")

    def save_temp_results(self, results, creator_name):
        """Save results to a temporary CSV for learning"""
        temp_dir = os.path.join(os.getcwd(), "temp_results")
        os.makedirs(temp_dir, exist_ok=True)

        temp_file = os.path.join(temp_dir,
                                 f"{creator_name.replace(' ', '_')}_temp_{datetime.now().strftime('%Y%m%d_%H%M%S')}.csv")

        if not results:
            return None

        df = pd.DataFrame(results)
        df.to_csv(temp_file, index=False)
        print(f"✅ Saved temp results to {temp_file} for keyword learning")

        return temp_file
